Mathlib/Data/Matrix/Notation.lean

```diff
@@ -0,0 +1,503 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anne Baanen, Eric Wieser
+
+! This file was ported from Lean 3 source module data.matrix.notation
+! leanprover-community/mathlib commit 3e068ece210655b7b9a9477c3aff38a492400aa1
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Matrix.Basic
+import Mathlib.Data.Fin.VecNotation
+import Mathlib.Tactic.FinCases
+import Mathlib.Tactic.ToExpr
+import Mathlib.Algebra.BigOperators.Fin
+import Qq
+
+/-!
+# Matrix and vector notation
+
+This file includes `simp` lemmas for applying operations in `Data.Matrix.Basic` to values built out
+of the matrix notation `![a, b] = vecCons a (vecCons b vecEmpty)` defined in
+`Data.Fin.VecNotation`.
+
+This also provides the new notation `!![a, b; c, d] = Matrix.of ![![a, b], ![c, d]]`.
+This notation also works for empty matrices; `!![,,,] : Matrix (Fin 0) (Fin 3)` and
+`!![;;;] : Matrix (Fin 3) (Fin 0)`.
+
+## Implementation notes
+
+The `simp` lemmas require that one of the arguments is of the form `vecCons _ _`.
+This ensures `simp` works with entries only when (some) entries are already given.
+In other words, this notation will only appear in the output of `simp` if it
+already appears in the input.
+
+## Notations
+
+This file provide notation `!![a, b; c, d]` for matrices, which corresponds to
+`Matrix.of ![![a, b], ![c, d]]`.
+Note that in lean 4 the pretty-printer will not show `!!` notation, instead showing the version
+with `of ![![...]]`.
+
+## Examples
+
+Examples of usage can be found in the `test/matrix.lean` file.
+-/
+
+namespace Matrix
+
+universe u uₘ uₙ uₒ
+
+variable {α : Type u} {o n m : ℕ} {m' : Type uₘ} {n' : Type uₙ} {o' : Type uₒ}
+
+open Matrix
+
+section toExpr
+open Lean
+open Qq
+
+/-- Matrices can be reflected whenever their entries can. We insert a `Matrix.of` to
+prevent immediate decay to a function. -/
+protected instance toExpr [ToLevel.{u}] [ToLevel.{uₘ}] [ToLevel.{uₙ}]
+    [Lean.ToExpr α] [Lean.ToExpr m'] [Lean.ToExpr n'] [Lean.ToExpr (m' → n' → α)] :
+    Lean.ToExpr (Matrix m' n' α) where
+  toTypeExpr :=
+    have eα : Q(Type $(toLevel.{u})) := toTypeExpr α
+    have em' : Q(Type $(toLevel.{uₘ})) := toTypeExpr m'
+    have en' : Q(Type $(toLevel.{uₙ})) := toTypeExpr n'
+    q(Matrix $eα $em' $en')
+  toExpr M :=
+    have eα : Q(Type $(toLevel.{u})) := toTypeExpr α
+    have em' : Q(Type $(toLevel.{uₘ})) := toTypeExpr m'
+    have en' : Q(Type $(toLevel.{uₙ})) := toTypeExpr n'
+    have eM : Q($em' → $en' → $eα) := toExpr (show m' → n' → α from M)
+    q(Matrix.of $eM)
+#align matrix.matrix.reflect Matrix.toExpr
+
+end toExpr
+
+section Parser
+open Lean Elab Term Macro TSyntax
+
+syntax "!![" sepBy1(sepBy1(term, ",", ", ", allowTrailingSep),
+              ";", "; ", allowTrailingSep) "]" : term
+syntax "!![" ";"+ "]" : term
+syntax "!![" ","+ "]" : term
+syntax "!![" "]" : term
```

Comment:
This syntax differs from #3429, so it will need another mathport bump if not reverted

Reply:
I synchronized syntaxes with `Mathport.Syntax`.